### PR TITLE
Try to find and verify existing OIDC providers before we try to create a new one

### PR DESCRIPTION
**Pull request by codablock**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
When moving clusters between management clusters, ControlPlane.Status.OIDCProvider.ARN
is lost. The new management cluster must then pickup the already existing
cluster, as otherwise it tries to create the same provider again and then
fails.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Checklist**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests
- [ ] adds or updates e2e tests

**Release note**:

```release-note
Try to find and verify existing OIDC providers before we try to create a new one
```

### CODE REVIEW

**Comment by richardcase** [comment]
Shame we stored this in the status in the original implementation. 

@codablock - do you see any potential issues (i.e. edge cases) with trying to to automatically find the oidc provider?

**Comment by codablock** [comment]
@richardcase I can't imagine any. I was first thinking that I could also try to update an existing OIDC provider's thumbprint ind client IDs, but then decided that I should not touch them as we can't know for sure WHY they would differ from the expected values...if they do, there is probably a good reason (e.g. manually created provider, for whatever reason).

**Comment by codablock** [comment]
@richardcase Any update on how to proceed with this PR?

**Comment by richardcase** [comment]
/lifecycle frozen

This looks good to me. I can't imagine any issues with this myself. @sedefsavas - what do you think?

**Comment by richardcase** [comment]
@codablock - this looks good to me. Do you think there is anyway we can add a test around this functionality?

**Comment by codablock** [comment]
Unfortunately I'm a bit overloaded atm and won't find time to add tests.

**Comment by Jacobious52** [comment]
bump. This one is fairly important for using workers across multiple managers

**Comment by Skarlso** [comment]
I'll take this.

**Comment by Skarlso** [comment]
To pick up:
- Test the code, and see if it works, if not, fix anything that might remain
- See if I can write some tests around this

**Comment by Skarlso** [comment]
This looks okay so far. I'll have a think about the security concerns such as, can this be abused to find an OIDC provider that you shouldn't have...

File: pkg/cloud/services/eks/iam/iam.go [hunk 1]
```diff
@@ -434,6 +435,44 @@ func (s *IAMService) CreateOIDCProvider(cluster *eks.Cluster) (string, error) {
...
+	thumbprint, err := fetchRootCAThumbprint(issuerURL.String())
+	if err != nil {
+		return "", err
+	}
+	output, err := s.IAMClient.ListOpenIDConnectProviders(&iam.ListOpenIDConnectProvidersInput{})
+	if err != nil {
+		return "", errors.Wrap(err, "error listing providers")
+	}
+	for _, r := range output.OpenIDConnectProviderList {
+		provider, err := s.IAMClient.GetOpenIDConnectProvider(&iam.GetOpenIDConnectProviderInput{OpenIDConnectProviderArn: r.Arn})
```
Comment (Skarlso) [inline pkg/cloud/services/eks/iam/iam.go:458]: FYI, this will increase the number of API calls we already have trouble managing. 🤔

Comment (Skarlso) [reply pkg/cloud/services/eks/iam/iam.go:458]: What if we moved out the arn field from under status?

Comment (Skarlso) [reply pkg/cloud/services/eks/iam/iam.go:458]: @richardcase @sedefsavas For any insight? :)

**Comment by Skarlso** [comment]
@codablock Hello 👋 

Would you mind giving me permission to push into this PR, please? :) 

Alternatively, I can open a new one under my name if you don't want to give me access. :D I understand. :)

**Comment by codablock** [comment]
@Skarlso as I don't know of a better way, I've just added you as a contributor to my fork 🙈

**Comment by Skarlso** [comment]
@codablock Sorry, I still don't have push permission. :( I'll continue in my own fork and reference your work. :)

**Comment by Skarlso** [comment]
@codablock I was being stupid. :D I didn't accept the invite yet. :DDDD

Please feel free to remove me once I'm done with this. :) Thank you!

**Comment by Skarlso** [comment]
This is a _nightmare_ to test. :D

**Comment by Skarlso** [comment]
Yesss!! Almost there. :D now trying to reconcile the LAST HURDLE. :D This part:

```go
	if err := s.reconcileTrustPolicy(); err != nil {
		return errors.Wrap(err, "failed to reconcile trust policy in workload cluster")
	}
```

:D

**Comment by Skarlso** [comment]
OH. MY. GODS.

```
I0929 22:29:41.349199   66964 oidc.go:54] "Reconciling EKS OIDC Provider" cluster-name=0xc000200960
--- PASS: TestOIDCReconcile (0.05s)
    --- PASS: TestOIDCReconcile/cluster_create_with_no_OIDC_provider_present_yet_should_create_one (0.05s)
PASS
```

**Comment by Skarlso** [comment]
Hah! I found a bug! :D :D Testing++++++

**Comment by Skarlso** [comment]
@richardcase @codablock This was DIFFICULT not gonna lie. :D But I added a test. :) And I found a bug where this was being done:

```go
		if fmt.Sprintf("https://%s", *provider.Url) != issuerURL.String() {
			continue
		}
```

But the provider url MUST contain `https` already anyways so this errored. Now, was there a reason for this being there?
I can add a thing like, check if https is there and if not, append it. But the specification is telling me there should be https on the retrieved provider always. 🤷

**Comment by Skarlso** [comment]
>TLS handshake error from 127.0.0.1:36514: remote error: tls: bad certificate

Sh*t. I was afraid of that. :/

**Comment by Skarlso** [comment]
Dang it. This won't work. :( I have to find a different approach and mock the http client completely.

File: pkg/cloud/services/eks/iam/iam.go [hunk 2]
```diff
@@ -434,8 +436,53 @@ func (s *IAMService) CreateOIDCProvider(cluster *eks.Cluster) (string, error) {
...
+		if len(provider.ClientIDList) != 1 || *provider.ClientIDList[0] != stsAWSAudience {
+			return "", errors.Wrap(err, "found provider with matching issuerURL but with non-matching clientID")
+		}
+		return *r.Arn, nil
+	}
+	return "", nil
+}
+
+func fetchRootCAThumbprint(issuerURL string, client *http.Client) (string, error) {
+	// needed to appease noctx.
```
Comment (Skarlso) [inline pkg/cloud/services/eks/iam/iam.go:479]: "Funny" how noctx never noticed "http.Get" but noticed client.Get immediately. Tsk, tsk, tsk.

**Comment by Skarlso** [comment]
This is now ready for review. :)

**Comment by Skarlso** [comment]
ping @Ankitasw as Richard is OOO. :))

File: pkg/cloud/services/eks/iam/iam.go (hunk 1 above)
Comment (richardcase) [reply pkg/cloud/services/eks/iam/iam.go:459]: We have taken this approach with other pieces of data. I would see no problem with the API version bump in moving this to the spec.

Comment (richardcase) [reply pkg/cloud/services/eks/iam/iam.go:459]: @sedefsavas - wdyt?

**Comment by Skarlso** [comment]
/test pull-cluster-api-provider-aws-e2e-eks

Hmm, hmm, unsure if this passes oidc part. Probably not. But at least it will test that EKS didn't break. :)

**Pull request merged**